community/modules/scripts/htcondor-install/files/autoscaler.py

```diff
@@ -180,47 +181,30 @@ def scale(self):
         else:
             self.zoneargs = {"zone": self.zone}
 
-        # Get total number of jobs in the queue that includes number of jobs
-        # waiting as well as number of jobs already assigned to nodes
-        queue_length_req = (
-            'condor_q -allusers -totals -format "%d " Jobs -format "%d " Idle -format "%d " Held'
-        )
-        queue_length_resp = os.popen(queue_length_req).read().split()
+        # Count jobs that are running or idle ("potentially runnable"); do not
+        # count held jobs or jobs in transitioning states
+        # https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.html#JobStatus
+        schedd = htcondor.Schedd()
+        job_attributes = [ "RequestCpus", "RequestMemory", "RequestGpus" ]
+        running_job_ads = schedd.query(constraint="JobStatus==2", projection=job_attributes)
+        idle_job_ads = schedd.query(constraint="JobStatus==1", projection=job_attributes)
 
-        if len(queue_length_resp) > 1:
-            queue = int(queue_length_resp[0])
-            idle_jobs = int(queue_length_resp[1])
-            on_hold_jobs = int(queue_length_resp[2])
-        else:
-            queue = 0
-            idle_jobs = 0
-            on_hold_jobs = 0
+        total_idle_request_cpus = sum(j["RequestCpus"] for j in idle_job_ads)
+        total_running_request_cpus = sum(j["RequestCpus"] for j in running_job_ads)
+        queue = total_idle_request_cpus + total_running_request_cpus
 
-        print("Total queue length: " + str(queue))
-        print("Idle jobs: " + str(idle_jobs))
-        print("Jobs on hold: " + str(on_hold_jobs))
```

Comment:
Why no longer report jobs on hold?

Reply:
The prior calculation consumed the stdout of a particular command that added up jobs in the queue. It did so in a manner that required subtracting off held jobs before calculating the autoscaling target. The new calculation computes the right target from the right data in the first place.

Jobs can go on hold deliberately or by HTCondor itself but the autoscaler isn't really the place to log that unless it's useful for autoscaling.